File: pipeline/fantasynfl/compute/records.py
"""All-time records across every season, computed directly from the database."""

from __future__ import annotations

import sqlite3
from dataclasses import dataclass

SINGLE_GAME_HIGH = "SINGLE_GAME_HIGH"
SINGLE_GAME_LOW = "SINGLE_GAME_LOW"
BIGGEST_WIN = "BIGGEST_WIN"
TOP_PLAYER_GAME = "TOP_PLAYER_GAME"
BEST_SEASON = "BEST_SEASON"
LONGEST_STREAK = "LONGEST_STREAK"

TOP_N = 5


@dataclass(frozen=True)
class RecordRow:
    category: str
    rank: int
    season_id: int | None
    team_id: int | None
    player_name: str | None
    value: float
    detail: str


_TEAM_GAMES_SQL = """
SELECT s.id AS season_id, s.year AS year, w.week_num AS week_num,
       m.home_team_id AS team_id, th.name AS team_name,
       m.home_score AS score, m.away_score AS opp_score, m.is_playoff AS is_playoff
FROM matchups m
JOIN weeks w ON w.id = m.week_id
JOIN seasons s ON s.id = w.season_id
JOIN teams th ON th.id = m.home_team_id
UNION ALL
SELECT s.id, s.year, w.week_num,
       m.away_team_id, ta.name,
       m.away_score, m.home_score, m.is_playoff
FROM matchups m
JOIN weeks w ON w.id = m.week_id
JOIN seasons s ON s.id = w.season_id
JOIN teams ta ON ta.id = m.away_team_id
"""


def compute_records(conn: sqlite3.Connection) -> list[RecordRow]:
    games = [dict(r) for r in conn.execute(_TEAM_GAMES_SQL).fetchall()]
    rows: list[RecordRow] = []
    rows += _single_game(games)
    rows += _biggest_win(games)
    rows += _top_player_game(conn)
    rows += _best_season(games)
    rows += _longest_streak(games)
    return rows


def _single_game(games: list[dict]) -> list[RecordRow]:
    high = sorted(games, key=lambda g: -g["score"])[:TOP_N]
    low = sorted(games, key=lambda g: g["score"])[:TOP_N]
    rows: list[RecordRow] = []
    for i, g in enumerate(high, 1):
        rows.append(
            RecordRow(
                SINGLE_GAME_HIGH,
                i,
                g["season_id"],
                g["team_id"],
                None,
                g["score"],
                f"{g['team_name']} · {g['score']:.1f} pts ({g['year']} W{g['week_num']})",
            )
        )
    for i, g in enumerate(low, 1):
        rows.append(
            RecordRow(
                SINGLE_GAME_LOW,
                i,
                g["season_id"],
                g["team_id"],
                None,
                g["score"],
                f"{g['team_name']} · {g['score']:.1f} pts ({g['year']} W{g['week_num']})",
            )
        )
    return rows


def _biggest_win(games: list[dict]) -> list[RecordRow]:
    wins = [g for g in games if g["score"] > g["opp_score"]]
    wins.sort(key=lambda g: -(g["score"] - g["opp_score"]))
    rows: list[RecordRow] = []
    for i, g in enumerate(wins[:TOP_N], 1):
        margin = g["score"] - g["opp_score"]
        rows.append(
            RecordRow(
                BIGGEST_WIN,
                i,
                g["season_id"],
                g["team_id"],
                None,
                margin,
                f"{g['team_name']} won by {margin:.1f} ({g['year']} W{g['week_num']})",
            )
        )
    return rows


def _top_player_game(conn: sqlite3.Connection) -> list[RecordRow]:
    sql = """
    SELECT r.player_name AS player_name, r.points AS points, r.position AS position,
           t.name AS team_name, s.year AS year, s.id AS season_id, t.id AS team_id
    FROM rosters r
    JOIN teams t ON t.id = r.team_id
    JOIN weeks w ON w.id = r.week_id
    JOIN seasons s ON s.id = w.season_id
    ORDER BY r.points DESC
    LIMIT ?
    """
    rows: list[RecordRow] = []
    for i, r in enumerate(conn.execute(sql, (TOP_N,)).fetchall(), 1):
        detail = f"{r['player_name']} ({r['position']}) · {r['points']:.1f} pts ({r['year']})"
        rows.append(
            RecordRow(
                TOP_PLAYER_GAME,
                i,
                r["season_id"],
                r["team_id"],
                r["player_name"],
                r["points"],
                detail,
            )
        )
    return rows


def _best_season(games: list[dict]) -> list[RecordRow]:
    agg: dict[tuple[int, int], dict] = {}
    for g in games:
        key = (g["season_id"], g["team_id"])
        a = agg.setdefault(key, {"wins": 0, "pf": 0.0, "name": g["team_name"], "year": g["year"]})
        a["pf"] += g["score"]
        if g["score"] > g["opp_score"]:
            a["wins"] += 1
    ranked = sorted(agg.items(), key=lambda kv: (kv[1]["wins"], kv[1]["pf"]), reverse=True)
    rows: list[RecordRow] = []
    for i, ((season_id, team_id), a) in enumerate(ranked[:TOP_N], 1):
        rows.append(
            RecordRow(
                BEST_SEASON,
                i,
                season_id,
                team_id,
                None,
                float(a["wins"]),
                f"{a['name']} · {a['wins']} wins ({a['year']})",
            )
        )
    return rows


def _longest_streak(games: list[dict]) -> list[RecordRow]:
    by_team: dict[tuple[int, int], list[dict]] = {}
    for g in games:
        by_team.setdefault((g["season_id"], g["team_id"]), []).append(g)

    best: list[tuple[int, int, str, int, int]] = []  # season, team, name, streak, year
    for (season_id, team_id), gs in by_team.items():
        gs.sort(key=lambda g: g["week_num"])
        streak = 0
        max_streak = 0
        name = gs[0]["team_name"]
        year = gs[0]["year"]
        for g in gs:
            if g["score"] > g["opp_score"]:
                streak += 1
                max_streak = max(max_streak, streak)
            else:
                streak = 0
        best.append((season_id, team_id, name, max_streak, year))

    best.sort(key=lambda x: -x[3])
    rows: list[RecordRow] = []
    for i, (season_id, team_id, name, streak, year) in enumerate(best[:TOP_N], 1):
        rows.append(
            RecordRow(
                LONGEST_STREAK,
                i,
                season_id,
                team_id,
                None,
                float(streak),
                f"{name} · {streak} straight wins ({year})",
            )
        )
    return rows

File: web/src/lib/queries/records.ts
import { db } from "../db";
import { getRevealState } from "../reveal";
import type { LuckRow, RecordRow, ShameItem } from "../types";
import { getRecapLuck } from "./recap-data";
import { getMaxRegularWeek } from "./seasons";
import { maskedTeamAbbrev, maskedTeamName } from "./shared";

interface ShameMatchupRow {
  id: number;
  week_num: number;
  home_team_id: number;
  away_team_id: number;
  home_score: number;
  away_score: number;
  winner_team_id: number | null;
  hname: string;
  habb: string;
  hcolor: string;
  aname: string;
  aabb: string;
  acolor: string;
  h_owner_alias_num: number | null;
  a_owner_alias_num: number | null;
}

export async function getRecords(): Promise<RecordRow[]> {
  const rows = db
    .prepare(
      `SELECT r.*, t.name AS record_team_name, o.alias_num AS owner_alias_num
       FROM records r
       LEFT JOIN teams t ON t.id = r.team_id
       LEFT JOIN owners o ON o.id = t.owner_id
       ORDER BY r.category, r.rank`
    )
    .all() as (RecordRow & { record_team_name: string | null })[];

  const revealed = await getRevealState();
  if (revealed) return rows;

  return rows.map((r) => {
    const { record_team_name, ...rest } = r;
    if (record_team_name === null) return rest;
    const detail =
      rest.detail !== null && rest.detail.startsWith(record_team_name)
        ? maskedTeamName(r.owner_alias_num) + rest.detail.slice(record_team_name.length)
        : rest.detail;
    return { ...rest, detail };
  });
}

export async function getShameData(seasonId: number): Promise<ShameItem[]> {
  const revealed = await getRevealState();
  const rawMatchups = db
    .prepare(
      `SELECT m.id, w.week_num, m.home_team_id, m.away_team_id, m.home_score, m.away_score,
              m.winner_team_id, th.name hname, th.abbrev habb, th.color hcolor,
              oh.alias_num AS h_owner_alias_num,
              ta.name aname, ta.abbrev aabb, ta.color acolor,
              oa.alias_num AS a_owner_alias_num
       FROM matchups m JOIN weeks w ON w.id = m.week_id
       JOIN teams th ON th.id = m.home_team_id
       LEFT JOIN owners oh ON oh.id = th.owner_id
       JOIN teams ta ON ta.id = m.away_team_id
       LEFT JOIN owners oa ON oa.id = ta.owner_id
       WHERE w.season_id = ?
       ORDER BY w.week_num`
    )
    .all(seasonId) as ShameMatchupRow[];

  const matchups = revealed
    ? rawMatchups
    : rawMatchups.map((m) => ({
        ...m,
        hname: maskedTeamName(m.h_owner_alias_num),
        habb: maskedTeamAbbrev(m.h_owner_alias_num),
        aname: maskedTeamName(m.a_owner_alias_num),
        aabb: maskedTeamAbbrev(m.a_owner_alias_num),
      }));

  if (matchups.length === 0) return [];

  const teamInfo = new Map<number, { name: string; abbrev: string; color: string }>();
  for (const m of matchups) {
    teamInfo.set(m.home_team_id, { name: m.hname, abbrev: m.habb, color: m.hcolor });
    teamInfo.set(m.away_team_id, { name: m.aname, abbrev: m.aabb, color: m.acolor });
  }

  const items: ShameItem[] = [];

  let biggestLoss: { margin: number; teamId: number; week: number; oppName: string } | null = null;
  let lowestScore: { score: number; teamId: number; week: number } | null = null;
  let cheapestWin: { score: number; teamId: number; week: number; oppName: string } | null = null;

  for (const m of matchups) {
    const sides = [
      {
        teamId: m.home_team_id,
        score: m.home_score,
        oppName: m.aname,
        oppScore: m.away_score,
      },
      {
        teamId: m.away_team_id,
        score: m.away_score,
        oppName: m.hname,
        oppScore: m.home_score,
      },
    ];

    for (const s of sides) {
      const lossMargin = s.oppScore - s.score;
      if (lossMargin > 0 && (!biggestLoss || lossMargin > biggestLoss.margin)) {
        biggestLoss = { margin: lossMargin, teamId: s.teamId, week: m.week_num, oppName: s.oppName };
      }
      if (!lowestScore || s.score < lowestScore.score) {
        lowestScore = { score: s.score, teamId: s.teamId, week: m.week_num };
      }
      if (m.winner_team_id === s.teamId && (!cheapestWin || s.score < cheapestWin.score)) {
        cheapestWin = { score: s.score, teamId: s.teamId, week: m.week_num, oppName: s.oppName };
      }
    }
  }

  if (biggestLoss) {
    const t = teamInfo.get(biggestLoss.teamId);
    items.push({
      kind: "BIGGEST_LOSS",
      label: "Biggest Loss",
      headline: `${t?.name ?? "Unknown"} lost by ${biggestLoss.margin.toFixed(1)}`,
      value: biggestLoss.margin,
      suffix: "pt loss",
      teamId: biggestLoss.teamId,
      teamName: t?.name ?? "Unknown",
      abbrev: t?.abbrev ?? "—",
      color: t?.color ?? "#888",
      detail: `Week ${biggestLoss.week} vs ${biggestLoss.oppName}`,
    });
  }

  if (lowestScore) {
    const t = teamInfo.get(lowestScore.teamId);
    items.push({
      kind: "LOWEST_SCORE",
      label: "Lowest Score",
      headline: `${t?.name ?? "Unknown"} managed just ${lowestScore.score.toFixed(1)} pts`,
      value: lowestScore.score,
      suffix: "pts",
      teamId: lowestScore.teamId,
      teamName: t?.name ?? "Unknown",
      abbrev: t?.abbrev ?? "—",
      color: t?.color ?? "#888",
      detail: `Week ${lowestScore.week}`,
    });
  }

  let longestSkid: { len: number; teamId: number } | null = null;
  for (const tid of teamInfo.keys()) {
    const games = matchups.filter((m) => m.home_team_id === tid || m.away_team_id === tid);
    let current = 0;
    let best = 0;
    for (const g of games) {
      if (g.winner_team_id === null) {
        current = 0;
        continue;
      }
      if (g.winner_team_id !== tid) {
        current++;
        best = Math.max(best, current);
      } else {
        current = 0;
      }
    }
    if (best > 0 && (!longestSkid || best > longestSkid.len)) {
      longestSkid = { len: best, teamId: tid };
    }
  }

  if (longestSkid) {
    const t = teamInfo.get(longestSkid.teamId);
    items.push({
      kind: "LONGEST_LOSING_STREAK",
      label: "Longest Losing Streak",
      headline: `${t?.name ?? "Unknown"} lost ${longestSkid.len} straight`,
      value: longestSkid.len,
      suffix: "straight",
      teamId: longestSkid.teamId,
      teamName: t?.name ?? "Unknown",
      abbrev: t?.abbrev ?? "—",
      color: t?.color ?? "#888",
      detail: "the league's longest skid",
    });
  }

  const luck = await getRecapLuck(seasonId, getMaxRegularWeek(seasonId));
  const unluckiest = luck.reduce<LuckRow | null>(
    (min, l) => (min === null || l.luck_score < min.luck_score ? l : min),
    null
  );
  if (unluckiest && unluckiest.luck_score < 0) {
    const robbed = Math.abs(unluckiest.luck_score);
    items.push({
      kind: "UNLUCKIEST",
      label: "Unluckiest",
      headline: `${unluckiest.name} was robbed ${robbed.toFixed(1)} wins`,
      value: robbed,
      suffix: "wins robbed",
      teamId: unluckiest.id,
      teamName: unluckiest.name,
      abbrev: teamInfo.get(unluckiest.id)?.abbrev ?? "—",
      color: unluckiest.color,
      detail: `expected ${unluckiest.expected_wins.toFixed(1)} wins, got ${unluckiest.actual_wins.toFixed(1)}`,
    });
  }

  if (cheapestWin) {
    const t = teamInfo.get(cheapestWin.teamId);
    items.push({
      kind: "CHEAPEST_WIN",
      label: "Cheapest Win",
      headline: `${t?.name ?? "Unknown"} won with just ${cheapestWin.score.toFixed(1)} pts`,
      value: cheapestWin.score,
      suffix: "pts",
      teamId: cheapestWin.teamId,
      teamName: t?.name ?? "Unknown",
      abbrev: t?.abbrev ?? "—",
      color: t?.color ?? "#888",
      detail: `Week ${cheapestWin.week} vs ${cheapestWin.oppName}`,
    });
  }

  return items;
}
